### PR TITLE
[meta] Enable PWA for website

### PR DESCRIPTION
**Pull request by hslee2008**
# Changes
There are many websites using Progressive Web App technology (Docusaurus) - React Native, Jest e.t.c.
I also think using Progressive Web App is very benefitial as it provides accessories for the users.
Using [Docusaurus plugin-pwa](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-pwa).

# Results Desktop
tauri_install
tauri_options
tauri_windows

# Results Mobile

### CODE REVIEW

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]: I don't know much about PWAs, but do you understand the strategies for [offline mode activation strategies](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-pwa#offlinemodeactivationstrategies)? I'm curious of the caution annotation in that section. `appInstalled` and `standalone` that you have sound like they're opt-in so fine. I read `queryString` as something that maybe we don't want as it's used for debugging? `saveData` seems fine.

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]: Do you know if it's possible to toggle these based on colour mode? Or to use a css variable instead? If we can use a css variable then `var(--ifm-color-primary)` would be nice to try

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]: Same comment as above about if we can change colours dynamically

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]: This image should be 180x180 pixels per Apple's HIG https://developer.apple.com/design/human-interface-guidelines/ios/icons-and-images/app-icon/

Comment (lorenzolewis) [reply docusaurus.config.js]: Should also be a solid background, not transparent

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]: I'm not sure about sizes on this one, does it need to be larger?

File: package.json
Comment (lorenzolewis) [inline package.json]: Remove the carot (^) from here, our dep checker will do it in another PR as soon as this is merged if not ;)

File: static/manifest.json
Comment (lorenzolewis) [inline static/manifest.json]: See note in config about image pixelation

File: static/manifest.json
Comment (lorenzolewis) [inline static/manifest.json]: Same comment on colour, can we use a CSS variable?

File: static/manifest.json
Comment (lorenzolewis) [inline static/manifest.json]: I like the idea of the shortcuts, but I'm worried about urls changing in the future (especially since they now have version numbers in them) and breaking. I'd opt to not have any shortcuts.

**Review by lorenzolewis — CHANGES_REQUESTED** [review]
I like the direction of this and don't see a downside. At first was worried about adding build time but only increased CI build time by like 3 seconds.

For the notes on images, let me know if you'd like me to help on that and I can throw together the proper images very quickly.

File: docusaurus.config.js
Comment (hslee2008) [reply docusaurus.config.js]: I think it needs to be 144x144. I saw some example code for msapplication-TileImage and all of them had the size of 144x144.

**Comment by hslee2008** [comment]
> I like the direction of this and don't see a downside. At first was worried about adding build time but only increased CI build time by like 3 seconds.
> 
> For the notes on images, let me know if you'd like me to help on that and I can throw together the proper images very quickly.

Thanks for the suggestion.

I don't think using CSS variables in manifest.json is possible.

docusaurus.io just used a theme color that is both readable with white/black text inside.

Same for background color.

**Comment by lorenzolewis** [comment]
iOS Icon
Can you use this file for the `apple-touch-icon` (will match up to the default Tauri icon on macOS)? Keep the current size because technically 1024x1024 is the largest iOS app size you export with an app, so should keep us future-proof.

**Comment by lorenzolewis** [comment]
Tauri Logo
And this is the vector for the logo if you'd like to re-render the other files so they aren't pixelated

**Comment by lorenzolewis** [comment]
I'd recommend having a margin around the edges of the files the width of one of the blue/yellow circles in the logo

**Comment by hslee2008** [comment]
Having trouble with generating icons (they are pixelated).
Could you upload the images (png) 144 and 512 (px)?.

> I'd recommend having a margin around the edges of the files the width of one of the blue/yellow circles in the logo

Do you mean the maskable images with margins around to make them maskable?

**Comment by lorenzolewis** [comment]
I'll make the images and include the margins tomorrow and put them here.

**Comment by lorenzolewis** [comment]
favicon144
favicon512

Here are these two. What's the maskable image used for? I think the 512 image should be able to work for both of them now since it's transparent? 

If you're running Windows can you give a screenshot of what the app icon looks like on the start menu and taskbar?

**Comment by hslee2008** [comment]
Maskable as in masking the image with shapes.
I see there are paddings around the logo in 512x512.
That works.
Thank you.

**Comment by lorenzolewis** [comment]
Just noticing that static/img/favicon-32x32.png is actually 31 pixels instead of 32 now. I think the original file that was there is fine (and the right size) but seems to be modified. 

Then lastly moving all the files into `static/meta` (images and manifests) just to keep that folder a bit more organised.

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]:
```suggestion
            content: '#0F0F0F',
```

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]:
```suggestion
            color: '#0F0F0F',
```

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]:
```suggestion
            content: '#0F0F0F',
```

File: static/manifest.json
Comment (lorenzolewis) [inline static/manifest.json]:
```suggestion
  "background_color": "#0F0F0F",
```

**Comment by lorenzolewis** [comment]
What was the outcome on separate manifests for dark/light mode? I'm putting in reviews for the theme colours but we have different ones depending on if it's dark or light, so was that able to be dynamic?

File: static/meta/manifest.json
Comment (lorenzolewis) [inline static/meta/manifest.json]:
```suggestion
  "background_color": "#0F0F0F",
```

File: static/meta/manifest.json
Comment (lorenzolewis) [inline static/meta/manifest.json]: Should this be set to meta/manifest.json now?

**Comment by hslee2008** [comment]
Light/Dark was a fail and did not work using media.
I think the best option is to use color that works in both light/dark theme as the background color.

**Comment by lorenzolewis** [comment]
Okay, I'll go through and suggest neutral colours on the theme then. Better to have that than a clash I think.

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]:
```suggestion
            content: '#0F0F0F',
```

File: static/meta/manifest.json
Comment (lorenzolewis) [inline static/meta/manifest.json]: Is this needed if it's pointing to the same image as the above lines?

File: static/meta/manifest.json
Comment (lorenzolewis) [inline static/meta/manifest.json]:
```suggestion
  "theme_color": "#0F0F0F",
```

File: docusaurus.config.js
Comment (lorenzolewis) [inline docusaurus.config.js]: Can this point to the one in the meta folder and then remove the one in the 'img' folder?

File: static/meta/manifest.json
Comment (lorenzolewis) [inline static/meta/manifest.json]: Can we move this file to meta and change the src here?

File: static/meta/manifest.json
Comment (lorenzolewis) [inline static/meta/manifest.json]: Same, move file, update source

**Comment by lorenzolewis** [comment]
Sorry for the million back and forths, but almost there!

**Comment by hslee2008** [comment]
> Sorry for the million back and forths, but almost there!

Is it okay if there are duplicate images in different directories?
Since favicon 16 and 32 is used in other files, I copied them to meta folder

**Comment by lorenzolewis** [comment]
Which other files are they used in? It would probably make sense to update where those files are pointing then if possible

**Comment by lorenzolewis** [comment]
Are you able to use yarn instead of npm to update yarn.lock? Right now it's changing all of the urls for every package which isn't ideal. If not then I can go back through and switch it over on my computer.

**Comment by hslee2008** [comment]
> Which other files are they used in? It would probably make sense to update where those files are pointing then if possible

They are used in docusaurus.config.js (169).
I will just update it.

**Comment by lorenzolewis** [comment]
I'll probably do some additional cleanup later for other assets, so just updating those for now is perfect

**Review by lorenzolewis — APPROVED** [review]

**Comment by lorenzolewis** [comment]
Looks good, but I want to get a Netlify deploy preview up and running first before merging. Not sure why it's paused, so I'll give a shout to the team over in Discord

**Comment by lorenzolewis** [comment]
Thanks for the netlify nudge @Laegel ;)

**Comment by lorenzolewis** [comment]
image

I'm going to tweak this image very slightly because of the corner artefacts

**Comment by lorenzolewis** [comment]
iOS Icon

Could you swap it to this one please?

**Comment by hslee2008** [comment]
Not sure why but there is an error in android.

**Comment by lorenzolewis** [comment]
When you launch it looks like it's trying to go to `/meta/`

File: static/meta/manifest.json
Comment (lorenzolewis) [inline static/meta/manifest.json]:
```suggestion
  "start_url": "/",
```

Comment (lorenzolewis) [reply static/meta/manifest.json]: I wonder if this will fix it

**Comment by lorenzolewis** [comment]
Need another approval on Netlify from the gods because we're close to our CI minutes @Laegel

**Comment by lorenzolewis** [comment]
Fixed for iOS, how about Android?

**Review by Laegel — APPROVED** [review]

**Comment by lorenzolewis** [comment]
@HyunseungLee-Travis can you verify you commits and then push some test again (just add a newline or a comment somewhere)? We're thinking that's why the previews stopped (and would be good practice to have all commits verified).

**Comment by hslee2008** [comment]
Works in Android.

**Comment by lorenzolewis** [comment]
Awesome job, thank you very much!

**Pull request merged**
